File: forms/neurodiversity-adjustment-request/front-end-with-svelte/src/lib/engine/impact-rules.ts
import type { NeurodiversityAdjustmentRequest, ImpactBand, FiredRule } from './types';

/** Impact bands in ascending severity; the highest firing rule wins. */
const ORDER: ImpactBand[] = ['ok', 'caution', 'high-risk'];

function raise(current: ImpactBand, next: ImpactBand): ImpactBand {
	return ORDER.indexOf(next) > ORDER.indexOf(current) ? next : current;
}

/**
 * Axis B — impact / wellbeing risk.
 *
 * Escalation ladder (ok → caution → high-risk). Every firing rule raises the
 * band to at least its level (max wins) and is recorded in the audit trail. A
 * worker at risk of sickness absence / burnout, or reporting severe current
 * impact, drives the band to high-risk. The least-alarming band is chosen only
 * when no rule fires.
 */
export function gradeImpact(r: NeurodiversityAdjustmentRequest): {
	impactBand: ImpactBand;
	firedRules: FiredRule[];
} {
	const firedRules: FiredRule[] = [];
	let band: ImpactBand = 'ok';

	if (r.atRiskOfAbsence) {
		firedRules.push({
			ruleId: 'R-IMPACT-ABSENCE-RISK',
			axis: 'impact',
			category: 'absence-risk',
			description:
				'Worker at risk of sickness absence or burnout without adjustments — act promptly.'
		});
		band = raise(band, 'high-risk');
	}

	if (r.currentImpact === 'severe') {
		firedRules.push({
			ruleId: 'R-IMPACT-SEVERE',
			axis: 'impact',
			category: 'severe-impact',
			description: 'Severe current impact on work and wellbeing.'
		});
		band = raise(band, 'high-risk');
	}

	if (r.currentImpact === 'high') {
		firedRules.push({
			ruleId: 'R-IMPACT-HIGH',
			axis: 'impact',
			category: 'high-impact',
			description: 'High current impact on work and wellbeing.'
		});
		band = raise(band, 'caution');
	}

	if (r.difficultyBurnoutWellbeing) {
		firedRules.push({
			ruleId: 'R-IMPACT-BURNOUT',
			axis: 'impact',
			category: 'burnout',
			description: 'Fatigue / burnout difficulty reported.'
		});
		band = raise(band, 'caution');
	}

	if (r.currentImpact === 'moderate') {
		firedRules.push({
			ruleId: 'R-IMPACT-MODERATE',
			axis: 'impact',
			category: 'moderate-impact',
			description: 'Moderate current impact on work and wellbeing.'
		});
		band = raise(band, 'caution');
	}

	if (firedRules.length === 0) {
		firedRules.push({
			ruleId: 'R-IMPACT-OK',
			axis: 'impact',
			category: 'no-risk',
			description: 'No wellbeing risk detected from the impact screen.'
		});
	}

	return { impactBand: band, firedRules };
}

File: forms/neurodiversity-adjustment-request/back-end-with-loco/src/engine/impact_rules.rs
//! Axis B — impact / wellbeing risk.
//!
//! Ported from `front-end-with-svelte/src/lib/engine/impact-rules.ts`. The band
//! escalates on the ladder `ok → caution → high-risk`; every firing rule raises
//! the band to at least its level (max wins) and all firing rules are collected.
//! The least-alarming band is chosen only when no rule fires.

use super::types::{FiredRule, NeurodiversityAdjustmentRequest};

/// The result of grading axis B.
pub struct Impact {
    /// The impact band (`ok` / `caution` / `high-risk`).
    pub impact_band: String,
    /// The audit-trail rules that fired.
    pub fired_rules: Vec<FiredRule>,
}

fn band_rank(band: &str) -> u8 {
    match band {
        "caution" => 1,
        "high-risk" => 2,
        _ => 0,
    }
}

/// Grade axis B — impact / wellbeing risk. Collect every firing rule and raise
/// the band to the highest level any of them implies.
#[must_use]
pub fn grade_impact(r: &NeurodiversityAdjustmentRequest) -> Impact {
    // (target-band, rule) pairs, evaluated in spec order.
    let mut candidates: Vec<(&str, FiredRule)> = Vec::new();

    if r.at_risk_of_absence {
        candidates.push((
            "high-risk",
            FiredRule::new(
                "R-IMPACT-ABSENCE-RISK",
                "impact",
                "absence-risk",
                "Worker at risk of sickness absence or burnout without adjustments — act promptly.",
            ),
        ));
    }
    if r.current_impact == "severe" {
        candidates.push((
            "high-risk",
            FiredRule::new(
                "R-IMPACT-SEVERE",
                "impact",
                "severe",
                "Severe current impact on work and wellbeing.",
            ),
        ));
    }
    if r.current_impact == "high" {
        candidates.push((
            "caution",
            FiredRule::new(
                "R-IMPACT-HIGH",
                "impact",
                "high",
                "High current impact on work and wellbeing.",
            ),
        ));
    }
    if r.difficulty_burnout_wellbeing {
        candidates.push((
            "caution",
            FiredRule::new(
                "R-IMPACT-BURNOUT",
                "impact",
                "burnout",
                "Fatigue / burnout difficulty reported.",
            ),
        ));
    }
    if r.current_impact == "moderate" {
        candidates.push((
            "caution",
            FiredRule::new(
                "R-IMPACT-MODERATE",
                "impact",
                "moderate",
                "Moderate current impact on work and wellbeing.",
            ),
        ));
    }

    if candidates.is_empty() {
        return Impact {
            impact_band: "ok".to_string(),
            fired_rules: vec![FiredRule::new(
                "R-IMPACT-OK",
                "impact",
                "ok",
                "No wellbeing risk detected from the impact screen.",
            )],
        };
    }

    let mut band = "ok".to_string();
    let mut fired_rules: Vec<FiredRule> = Vec::with_capacity(candidates.len());
    for (target, rule) in candidates {
        if band_rank(target) > band_rank(&band) {
            band = target.to_string();
        }
        fired_rules.push(rule);
    }

    Impact {
        impact_band: band,
        fired_rules,
    }
}
